Add `preview_datadog_agent_sampling ` (#6112)
apollographql/router: docs/source/configuration/telemetry/exporters/tracing/datadog.mdx

| `span_metrics`        | See [config](#span_metrics)         | A map of span names to boolean.         |

## Sampler configuration

When using Datadog you will need to take into consideration if you want to use the Datadog APM view or rely on OTLP metrics to gain insight into the Router's performance.
The Datadog APM view is driven by traces, and for this to be accurate 100% of requests must be sampled and sent to the Datadog agent.

Tracing is expensive both in terms of APM costs but also Router performance, and typically you will want to set the `sampler` to a low value ion production environments.
However, this will mean that the APM view will only show a small percentage of traces.

Datadog Agent sampling is a mode where ALL traces are sent to the Datadog agent, but only a percentage of them are forwarded to Datadog. This makes the APM view accurate while keeping costs low
at the cost of the Router having an effective sample rate of 100% under the hood.

Use the following guidelines on how to configure the `sampler` and `preview_datadog_agent_sampling` to get the desired behavior:

> **I want the APM view to show metrics for 100% of traffic, and I am OK with the performance impact on the Router.**

Reviewer: I like these bold, conversational-style "headings". So far in docs we haven't used these, but they're easy to understand. May reformat them in the future into lists.

```suggestion
**I want the APM view to show metrics for 100% of traffic, and I am OK with the performance impact on the router.**
```

Set `preview_datadog_agent_sampling` to `true` and adjust the `sampler` to the desired percentage of traces to be sent to Datadog.

```yaml title="router.yaml"
telemetry:
  exporters:
    tracing:
      common:
        # All requests will be traced and sent to the Datadog agent.
        # Only 10 percent of spans will be forwarded from the Datadog agent to Datadog.
        preview_datadog_agent_sampling: true
        sampler: 0.1
```

> **I want the Datadog agent to be in control of the percentage of traces sent to Datadog.**

Reviewer: ```suggestion
**I want the Datadog Agent to be in control of the percentage of traces sent to Datadog.**
```

> Use the Datadog agent `probabalistic_sampling` option sampler and set the `sampler` to `always_on` to allow the Datadog agent to control the sampling rate.

Reviewer: ```suggestion
Use the Datadog Agent's `probabalistic_sampling` option sampler and set the `sampler` to `always_on` to allow the agent to control the sampling rate.
```

Router config:
```yaml title="router.yaml"
telemetry:
  exporters:
    tracing:
      common:
        # All requests will be traced and sent to the Datadog agent.
        sampler: always_on
```

Datadog agent config:
```yaml
otlp_config:
  traces:
    probabilistic_sampling:
      # Only 10 percent of spans will be forwarded to Datadog
      sampling_percentage: 10
```

> **I want the most performance from the Router and am not concerned with the APM view. I use metrics and traces to monitor my application.**

Reviewer: ```suggestion
**I want the best performance from the router and I'm not concerned with the APM view. I use metrics and traces to monitor my application.**
```

```yaml title="router.yaml"
telemetry:
  exporters:
    tracing:
      common:
        # Only 10 percent of spans will be forwarded to the Datadog agent. Experiment to find a value that is good for you!
        sampler: 0.1
```

If you are using the Datadog APM viw then you should set `preview_datadog_agent_sampling` to `true` and adjust the `sampler` to the desired percentage of traces to be sent to Datadog.

### `preview_datadog_agent_sampling` (default: `false`)

The Datadog APM view relies on traces to generate metrics. For this to be accurate 100% of requests must be sampled and sent to the Datadog agent.
To prevent ALL traces from then being sent to Datadog, you must set `preview_datadog_agent_sampling` to `true` and adjust the `sampler` to the desired percentage of traces to be sent to Datadog.

```yaml title="router.yaml"
telemetry:
  exporters:
    tracing:
      common:
        # Only 10 percent of spans will be forwarded from the Datadog agent to Datadog. Experiment to find a value that is good for you!
        preview_datadog_agent_sampling: true
        sampler: 0.1
```

> <Note>
> 
>     Using `preview_datadog_agent_sampling` will send ALL spans to the Datadog agent, but only the `sampler` percentage of them will be forwarded to Datadog. This means that your APM view will be correct at the cost of
>     the Router sending more spans to the Datadog agent. This will have an impact on the resource usage and performance of both the Router and Datadog agent.
> 
>     If you are OK with your APM view only showing a subset of traces, then you can leave `preview_datadog_agent_sampling` to `false`, however it is recommended to rely on OTLP metrics to gain insight into the Router's performance.
> 
> </Note>

Reviewer: Move this content out of the Note, while also copying the Notes from the overview here.

```suggestion
Using `preview_datadog_agent_sampling` will send _all_ spans to the Datadog Agent, but only the percentage of traces configured by the `sampler` will be forwarded to Datadog. This means that your APM view will be accurate, but it will incur performance and resource usage costs for both the router and Datadog Agent to send and receive all spans.

If your use case allows your APM view to show only a subset of traces, then you can set `preview_datadog_agent_sampling` to `false`. You should alternatively rely on OTLP metrics to gain insight into the router's performance.

<Note>
    
- The router doesn't support [`in-agent` ingestion control](https://docs.datadoghq.com/tracing/trace_pipeline/ingestion_mechanisms/?tab=java#in-the-agent).
    
- Configuring `traces_per_second` in the Datadog Agent will not dynamically adjust the router's sampling rate to meet the target rate.

</Note>
```
